src/ring_theory/tensor_product.lean

```diff
@@ -1,36 +1,321 @@
+# Tensor products and algebras
+
+Let `R` be a (semi)ring and `A` an `R`-algebra.
+In this file we:
+
+- Define the `A`-module structure on `A ⊗ M`, for an `R`-module `M`.
+- Define the `R`-algebra structure on `A ⊗ B`, for another `R`-algebra `B`.
+  and provide the structure isomorphisms
+  * `R ⊗[R] A ≃ₐ[R] A`
+  * `A ⊗[R] R ≃ₐ[R] A`
+  * `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+  The code for
+  * `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
+  is written and compiles, but takes longer than the `-T100000` time limit,
+  so is currently commented out.
 
-We construct the R-algebra structure on `A ⊗[R] B`, when `A` and `B` are both `R`-algebras,
-and provide the structure isomorphisms
+## Main declaration
 
-* `R ⊗[R] A ≃ₐ[R] A`
-* `A ⊗[R] R ≃ₐ[R] A`
-* `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+- `linear_map.base_change A f` is the `A`-linear map `A ⊗ f`, for an `R`-linear map `f`.
 
-The code for
-* `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
-is written and compiles, but takes longer than the `-T100000` time limit,
-so is currently commented out.
 -/
 
+open_locale tensor_product
+open tensor_product
+
+-- move this
 namespace algebra
 
-open_locale tensor_product
+variables {R A M N P : Type*}
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule R M] [semimodule A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [semimodule R N]
+
+variables (R M)
+
+/-- The `R`-algebra morphism `A → End (M)` corresponding to the `A`-module structure on `M`. -/
+def lsmul : A →ₐ[R] (M →ₗ[R] M) :=
+{ map_one' := by { ext m, exact one_smul A m },
+  map_mul' := by { intros a b, ext c, exact smul_assoc a b c },
+  map_zero' := by { ext m, exact zero_smul A m },
+  commutes' := by { intro r, ext m, exact algebra_map_smul A r m },
+  .. (show A →ₗ[R] M →ₗ[R] M, from linear_map.mk₂ R (•)
+  (λ x y z, add_smul x y z)
+  (λ c x y, smul_assoc c x y)
+  (λ x y z, smul_add x y z)
+  (λ c x y, smul_algebra_smul_comm c x y)) }
+
+@[simp] lemma lsmul_coe (a : A) : (lsmul R M a : M → M) = (•) a := rfl
+
+instance : is_scalar_tower R A A :=
+{ smul_assoc := λ r a b, by simp only [smul_eq_mul, smul_mul_assoc] }
+
+end algebra
+
+namespace tensor_product
+
+variables {R A M N P : Type*}
+
+section restrict_scalars
+variables [comm_semiring R] [comm_semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule A M]
+variables [add_comm_monoid N] [semimodule A N]
+
+variables (R A M N)
+
+@[priority 500]
+instance restrict_scalars.semimodule : semimodule R (M ⊗[A] N) :=
+show semimodule R (restrict_scalars R A (M ⊗ N)), from
+restrict_scalars.semimodule R A (M ⊗ N)
+
+instance restrict_scalars.is_scalar_tower : is_scalar_tower R A (M ⊗[A] N) :=
+show is_scalar_tower R A (restrict_scalars R A (M ⊗ N)), from
+restrict_scalars.is_scalar_tower R A (M ⊗ N)
+
+variables [semimodule R M] [semimodule R N]
+
+/- Check that we didn't introduce a diamond. -/
+example : tensor_product.restrict_scalars.semimodule R R M N =
```

Comment:
Should this be in the tests folder?

Reply:
I would like to keep it here. It doesn't take up a lot of spacetime, and it is an important check that people reading/using this file might be worried about.

Reply:
Resolving as the instance and thus this example have now been removed.